src/main/java/com/laytonsmith/core/functions/AbstractFunction.java

```diff
@@ -70,13 +71,26 @@ public Mixed execs(Target t, Environment env, Script parent, ParseTree... nodes)
 		return CVoid.VOID;
 	}
 
+	@Override
+	public FunctionSignatures getSignatures() {
+		return null;
+	}
+
 	/**
 	 * {@inheritDoc} By default, {@link CClassType#AUTO} is returned.
 	 */
 	@Override
 	public CClassType getReturnType(Target t, List<CClassType> argTypes,
 			List<Target> argTargets, Environment env, Set<ConfigCompileException> exceptions) {
-		return CClassType.AUTO; // No information is available about the return type.
+
+		// Match arguments to function signatures if available.
+		FunctionSignatures signatures = this.getSignatures();
```

Comment:
Probably worth it to cache this.

Reply:
I've created a cache getter that will come in convenient when getting signatures for documentation and reflective access as well.